internal/controller/impersonatorconfig/impersonator_config.go

```diff
@@ -521,37 +524,109 @@ func (c *impersonatorConfigController) ensureClusterIPServiceIsStopped(ctx conte
 	return utilerrors.FilterOut(err, k8serrors.IsNotFound)
 }
 
-func (c *impersonatorConfigController) createOrUpdateService(ctx context.Context, service *v1.Service) error {
-	log := c.infoLog.WithValues("serviceType", service.Spec.Type, "service", klog.KObj(service))
-	existing, err := c.servicesInformer.Lister().Services(c.namespace).Get(service.Name)
+func (c *impersonatorConfigController) createOrUpdateService(ctx context.Context, desiredService *v1.Service) error {
+	log := c.infoLog.WithValues("serviceType", desiredService.Spec.Type, "service", klog.KObj(desiredService))
+
+	// Prepare to remember which annotation keys were added from the CredentialIssuer spec, both for
+	// creates and for updates, in case someone removes a key from the spec in the future. We would like
+	// to be able to detect that the missing key means that we should remove the key. This is needed to
+	// differentiate it from a key that was added by another actor, which we should not remove.
+	// But don't bother recording the requested annotations if there were no annotations requested.
+	desiredAnnotationKeys := []string{}
+	for k := range desiredService.Annotations {
+		desiredAnnotationKeys = append(desiredAnnotationKeys, k)
+	}
+	if len(desiredAnnotationKeys) > 0 {
+		// Sort them since they come out of the map in no particular order.
+		sort.Strings(desiredAnnotationKeys)
+		keysJSONArray, err := json.Marshal(desiredAnnotationKeys)
+		if err != nil {
+			return err // This shouldn't really happen. We should always be able to marshal an array of strings.
+		}
+		if desiredService.Annotations == nil {
+			desiredService.Annotations = map[string]string{}
+		}
+		// Save the desired annotations to a bookkeeping annotation.
+		desiredService.Annotations[annotationKeysKey] = string(keysJSONArray)
+	}
+
+	// Get the Service from the informer, and create it if it does not already exist.
+	existingService, err := c.servicesInformer.Lister().Services(c.namespace).Get(desiredService.Name)
 	if k8serrors.IsNotFound(err) {
 		log.Info("creating service for impersonation proxy")
-		_, err := c.k8sClient.CoreV1().Services(c.namespace).Create(ctx, service, metav1.CreateOptions{})
+		_, err := c.k8sClient.CoreV1().Services(c.namespace).Create(ctx, desiredService, metav1.CreateOptions{})
 		return err
 	}
 	if err != nil {
 		return err
 	}
 
-	// Update only the specific fields that are meaningfully part of our desired state.
-	updated := existing.DeepCopy()
-	updated.ObjectMeta.Labels = service.ObjectMeta.Labels
-	updated.ObjectMeta.Annotations = service.ObjectMeta.Annotations
-	updated.Spec.LoadBalancerIP = service.Spec.LoadBalancerIP
-	updated.Spec.Type = service.Spec.Type
-	updated.Spec.Selector = service.Spec.Selector
+	// The Service already exists, so update only the specific fields that are meaningfully part of our desired state.
+	updatedService := existingService.DeepCopy()
+	updatedService.ObjectMeta.Labels = desiredService.ObjectMeta.Labels
```

Comment:
Probably, yes. I chatted with @mattmoyer about this yesterday and we were thinking that maybe we can defer that effort, since it wasn't the specific problem reported by the user and because we're not sure if we should keep the custom labels feature or not. 😄 But if we think its worth the effort then I could see if there's some way to make this merge code generic enough to handle both.